### PR TITLE
core: frontend: libs: Add MAVLink2Rest library

### PR DESCRIPTION
**Pull request by patrickelectric**
To test: patrickelectric/companion-core:mavlink2rest_ts
(you'll see MAV_TYPE_SUBMARINE at 10Hz)

Check console output, you'll see the attitude values. 

The code is the following:
```js
mavlink2rest.startListening('HEARTBEAT').setCallback((message) => {
  console.log(message.mavtype.type)
}).setFrequency(10)
```

### CODE REVIEW

File: core/frontend/src/libs/MAVLink2Rest.ts [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+  baseUrl: string;
+  // Dictionary mapping endpoints to websocket
+  endpoints: Dictionary<Endpoint> = {}
+  baseUrlCandidates: Array<string>
+
+  private static instance: Mavlink2RestManager;
+
+  private constructor () {
+    this.baseUrl = `${this.getWebsocketPrefix()}://${window.location.host}/mavlink2rest/ws/mavlink`
+    this.baseUrlCandidates = [this.baseUrl, 'ws://localhost:8088/ws/mavlink']
```
Comment (Williangalvani) [inline core/frontend/src/libs/MAVLink2Rest.ts:109]: what is up with the second candidate?
IIRC in the original code it was for local development, but I'm not sure if that still applies.

File: core/frontend/src/libs/MAVLink2Rest.ts [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    if (!Mavlink2RestManager.instance) {
+      Mavlink2RestManager.instance = new Mavlink2RestManager()
+    }
+    return Mavlink2RestManager.instance
+  }
+
+  /**
+   * Check for valid base URL from REST API
+   */
+  probeBaseUrlCandidates () {
```
Comment (Williangalvani) [inline core/frontend/src/libs/MAVLink2Rest.ts:127]: is this working to find the correct url? is it required still?

**Review by Williangalvani — COMMENTED** [review]
I got a couple of questions.

I like the code overall. it feels... familiar :eyes:

File: core/frontend/src/libs/MAVLink2Rest.ts (hunk 1 above)
Comment (patrickelectric) [reply core/frontend/src/libs/MAVLink2Rest.ts:109]: With that we could run the frontend and the service locally without running entire companion core

File: core/frontend/src/libs/MAVLink2Rest.ts (hunk 2 above)
Comment (patrickelectric) [reply core/frontend/src/libs/MAVLink2Rest.ts:127]: Check my previous comment

**Review by Williangalvani — APPROVED** [review]

**Comment by rafaellehmkuhl** [comment]
Code looks good.
Was wondering though if wouldn't it be better for it to be available as an external dependency, like the marvlink2rest itself?

**Comment by patrickelectric** [comment]
> Was wondering though if wouldn't it be better for it to be available as an external dependency, like the marvlink2rest itself?

How so ?

**Comment by rafaellehmkuhl** [comment]
> > Was wondering though if wouldn't it be better for it to be available as an external dependency, like the marvlink2rest itself?
> 
> How so ?

Publishing on NPM.

**Comment by patrickelectric** [comment]
> Publishing on NPM.

For the same reason that our python services are not in pypa, we are not going to use it outside this project and it's an unnecessary bureaucracy.

**Comment by rafaellehmkuhl** [comment]
I was not suggesting for bureaucracy reasons, but as a feature to your package, since mavlink2rest itself is already a library and probably it's users would benefit from having a javascript library available also.

If you prefer not to, it's ok.

**Comment by patrickelectric** [comment]
> I was not suggesting for bureaucracy reasons, but as a feature to your package, since mavlink2rest itself is already a library and probably it's users would benefit from having a javascript library available also.

Oh, good idea!
Sadly I'm not going to have time to do such thing, but maybe in a future where I could have typescript object definition for each supported mavlink message.

**Pull request merged**